### DISCUSSION TITLE
A way to get useForm return values in resolver?

### DISCUSSION THREAD

**Discussion by sebastianlp** [react-hook-form/react-hook-form#4247] [Q&A] [answered]
I didn't find a way to get `useForm` return values inside the resolver, it's possible? 

My use case is that I need to work with the `formState.dirtyFields` to make some dynamic validation inside my custom resolver but I don't know if this is possible. My first approach was to save `dirtyFields` on a local state and update it but that produce an extra re-render

**Comment by bluebill1049** [comment, marked as answer]
`resolver: (data) => data`

isn't `resolver` already return `data`?

**Reply by sebastianlp** [reply]
Actually yes but I'm creating an edit form so I already have data preloaded. I can add more context: 

I have two inputs (for simplicity but currently I'm using `useFieldArray` so I've a lot of this "tuple" inputs)
```js
A: 'a',
B: 'b'
```

I want to alert the user that `B` may need a change when `A` is dirty (or changed) and I was wondering if I can rely on `formState.dirtyFields.A` to accomplish that

**Reply by leapful** [reply]
@sebastianlp 

Yes, you can do it with formState.dirtyFields. Just remember to read the formState before evaluating its values.

For example:

```
// Read the values of formState first to subscribe it via Proxy
const { formState: { isDrity. dirtyFields } } = useForm();
```

NOTE: In case you want to perform validation with resolver or show warning on some fields when other fields changed to dirty, mode "onChange" or at least "onBlur" should be used because react-hook-form uses mode "onSubmit" by default, therefore "resolver" is not triggered and "formState" is not updated until end-user submits the form.

```
const { formState: { isDirty, dirtyFields } } = useForm({ 
  mode: 'onChange' ,
  resolver: async () => {
  // Evaluate isDirty and dirtyFields here
  }
})

```

**Reply by sebastianlp** [reply]
Thanks guys, @leapful this is what I finally did


```js
const customResolver = async (data, dirtyFields) => {
// run schema validation + using dirtyFields
}
const component = ({}) => {
  const {
    control,
    trigger,
    errors,
    handleSubmit,
    formState: {dirtyFields},
  } = useForm({
    defaultValues: {
      myArray: myArrayWithDefaultValues,
    },
    resolver: async data => {
      return customResolver(data, dirtyFields)
    },
    mode: 'onChange',
  })
```
